integration/gateway/gateway_test.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright IBM Corp All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package gateway
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/hyperledger/fabric-protos-go/gateway"
+	"github.com/hyperledger/fabric/integration/nwo"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/tedsuo/ifrit"
+	"google.golang.org/grpc"
+)
+
+var _ = Describe("GatewayService", func() {
+	var (
+		testDir       string
+		client        *docker.Client
+		network       *nwo.Network
+		process       ifrit.Process
+		orderer       *nwo.Orderer
+		org1Peer0     *nwo.Peer
+		org2Peer0     *nwo.Peer
+		conn          *grpc.ClientConn
+		gatewayClient gateway.GatewayClient
+	)
+
+	BeforeEach(func() {
+		var err error
+		testDir, err = ioutil.TempDir("", "gateway")
+		Expect(err).NotTo(HaveOccurred())
+
+		client, err = docker.NewClientFromEnv()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		if process != nil {
+			process.Signal(syscall.SIGTERM)
+			Eventually(process.Wait(), network.EventuallyTimeout).Should(Receive())
+		}
+		if network != nil {
+			network.Cleanup()
+		}
+		os.RemoveAll(testDir)
+	})
+
+	Describe("evaluate with result", func() {
+		BeforeEach(func() {
+			config := nwo.BasicEtcdRaft()
+			network = nwo.New(config, testDir, client, StartPort(), components)
+
+			network.GatewayEnabled = true
+
+			network.GenerateConfigTree()
+			network.Bootstrap()
+
+			networkRunner := network.NetworkGroupRunner()
+			process = ifrit.Invoke(networkRunner)
+			Eventually(process.Ready(), network.EventuallyTimeout).Should(BeClosed())
+
+			orderer = network.Orderer("orderer")
+			org1Peer0 = network.Peer("Org1", "peer0")
+			org2Peer0 = network.Peer("Org2", "peer0")
+
+			network.CreateAndJoinChannel(orderer, "testchannel")
+			network.UpdateChannelAnchors(orderer, "testchannel")
+			network.VerifyMembership(network.PeersWithChannel("testchannel"), "testchannel")
+
+			nwo.EnableCapabilities(network, "testchannel", "Application", "V2_0", orderer, org1Peer0, org2Peer0)
+
+			chaincodePath, err := filepath.Abs("../chaincode/gateway")
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
Can we please flatten this structure? We have a Describe/Describe/Context/It nesting for a single test.
There's also no symmetry between the before / after each implementations at the various levels.
There's also no apparent structure to the collection of the vars at the top level relative to where they are located.

Given we're dealing with one network, creating one client, to drive one scenario, I'd expect a single BeforeEach, a single AfterEach, and a single It (that has an implied context) .